src/windows/win_broker_os_interface.cpp

```diff
@@ -0,0 +1,177 @@
+/******************************************************************************
+ * Copyright 2019 ETC Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ ******************************************************************************
+ * This file is a part of RDMnetBroker. For more information, go to:
+ * https://github.com/ETCLabs/RDMnetBroker
+ *****************************************************************************/
+
+#include "win_broker_os_interface.h"
+
+#include <iostream>
+#include <memory>
+#include <Windows.h>
+#include <ShlObj.h>
+
+constexpr const WCHAR kRelativeConfFileName[] = L"\\ETC\\RDMnetBroker\\broker.conf";
+static const std::vector<std::wstring> kRelativeLogFilePath = {L"ETC", L"RDMnetBroker"};
+static const std::wstring kLogFileName = L"broker.log";
+
+std::unique_ptr<char[]> ConvertPathToUtf8(const std::wstring& conf_file_path)
+{
+  // Convert the path to UTF-8
+  int size_needed = WideCharToMultiByte(CP_UTF8, 0, conf_file_path.c_str(), -1, NULL, 0, NULL, NULL);
+  if (size_needed > 0)
+  {
+    auto buf = std::make_unique<char[]>(size_needed);
+    int convert_res = WideCharToMultiByte(CP_UTF8, 0, conf_file_path.c_str(), -1, buf.get(), size_needed, NULL, NULL);
+    if (convert_res > 0)
+    {
+      return buf;
+    }
+  }
+  return std::unique_ptr<char[]>();
+}
+
+WindowsBrokerOsInterface::WindowsBrokerOsInterface()
+{
+  PWSTR program_data_path;
+  HRESULT get_known_folder_res = SHGetKnownFolderPath(FOLDERID_ProgramData, 0, NULL, &program_data_path);
+  if (get_known_folder_res == S_OK)
+  {
+    program_data_path_ = program_data_path;
+    CoTaskMemFree(program_data_path);
+
+    // Build the log file path on top of it
+    log_file_path_ = program_data_path_;
+    for (const auto& intermediate_dir : kRelativeLogFilePath)
+    {
+      log_file_path_ += L"\\" + intermediate_dir;
+    }
+    log_file_path_ += L"\\" + kLogFileName;
+  }
+}
+
+WindowsBrokerOsInterface::~WindowsBrokerOsInterface()
+{
+  if (log_file_)
+    fclose(log_file_);
+}
+
+std::string WindowsBrokerOsInterface::GetLogFilePath() const
+{
+  if (log_file_path_.empty())
+    return std::string{};
+
+  auto convert_res = ConvertPathToUtf8(log_file_path_);
+  if (convert_res)
+    return convert_res.get();
+  else
+    return std::string{};
+}
+
+bool WindowsBrokerOsInterface::OpenLogFile()
+{
+  if (program_data_path_.empty())
```

Comment:
What happens if OpenLogFile is mistakenly called twice?

Reply:
The current answer is probably "_wfsopen() will return an error since the file is already open". But you raise a good point, I will add an assert() for this.